File: TS/系统/06．经济系统/01．杀敌金币平分/02．金币获取率加成.ts
/** @noSelfInFile */
/**
 * 杀敌金币平分系统 - 金币获取率加成
 *
 * 功能：处理金币获取率加成
 */

const { GOLD_RATE_THRESHOLD } = require("系统.06．经济系统.01．杀敌金币平分.00．常量定义") as typeof import("./00．常量定义");

const { YDUserDataGet } = require("lib.扩展函数.YDWE函数.01．YDUserData兼容") as {
  YDUserDataGet: (tableType: string, tableKey: any, attr: string, valueType: string) => any;
};

const { registerGoldGainCallback } = require("系统.06．经济系统.01．杀敌金币平分.01．核心功能") as {
  registerGoldGainCallback: (cb: (params: any) => any) => void;
};

// ==========================================================================================
// 工具函数
// ==========================================================================================

/** 获取玩家金币获取率 */
function getPlayerGoldRate(this: void, player: any): number {
  const rate = YDUserDataGet("player", player, "金币获取率", "real");
  return typeof rate === "number" ? rate : 0;
}

// ==========================================================================================
// 金币获取率回调
// ==========================================================================================

/**
 * 处理金币获取率加成
 */
function goldRateCallback(params: any): any {
  const { player, baseGold } = params;

  // 计算金币获取率加成
  const goldRate = getPlayerGoldRate(player);
  let finalGold = baseGold;

  if (goldRate >= GOLD_RATE_THRESHOLD) {
    finalGold = Math.floor(baseGold * (1 + goldRate));
  }

  return { ...params, finalGold };
}

registerGoldGainCallback(goldRateCallback);

export {};

File: src/系统/06．经济系统/01．杀敌金币平分/02．金币获取率加成.lua
local ____lualib = require("lualib_bundle")
local __TS__ObjectAssign = ____lualib.__TS__ObjectAssign
local ____exports = {}
local ____require_result_0 = require("系统.06．经济系统.01．杀敌金币平分.00．常量定义")
local GOLD_RATE_THRESHOLD = ____require_result_0.GOLD_RATE_THRESHOLD
local ____require_result_1 = require("lib.扩展函数.YDWE函数.01．YDUserData兼容")
local YDUserDataGet = ____require_result_1.YDUserDataGet
local ____require_result_2 = require("系统.06．经济系统.01．杀敌金币平分.01．核心功能")
local registerGoldGainCallback = ____require_result_2.registerGoldGainCallback
--- 获取玩家金币获取率
local function getPlayerGoldRate(player)
    local rate = YDUserDataGet(
        nil,
        "player",
        player,
        "金币获取率",
        "real"
    )
    return type(rate) == "number" and rate or 0
end
--- 处理金币获取率加成
local function goldRateCallback(params)
    local ____params_3 = params
    local player = ____params_3.player
    local baseGold = ____params_3.baseGold
    local goldRate = getPlayerGoldRate(player)
    local finalGold = baseGold
    if goldRate >= GOLD_RATE_THRESHOLD then
        finalGold = math.floor(baseGold * (1 + goldRate))
    end
    return __TS__ObjectAssign({}, params, {finalGold = finalGold})
end
registerGoldGainCallback(nil, goldRateCallback)
return ____exports
